### PR TITLE
gh-103092: Isolate socket

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-103092

### CODE REVIEW

File: Modules/socketmodule.c
Comment (kumaraditya303) [inline Modules/socketmodule.c]:
```suggestion
    int *atomic_flag_works = &state->sock_cloexec_works;
```

Nit: parens look weird to me.

File: Modules/socketmodule.c
Comment (kumaraditya303) [inline Modules/socketmodule.c]:
```suggestion
    int *atomic_flag_works = &state->sock_cloexec_works;
```

**Review by kumaraditya303 — APPROVED** [review]

File: Modules/socketmodule.c
Comment (erlend-aasland) [reply Modules/socketmodule.c]: Fair enough

**Comment by erlend-aasland** [comment]
There's some leaks that needs to be adressed:

```console
$ ./python.exe measure.py
before=46451, after=47809
before=47811, after=48549
before=48549, after=49287
before=49287, after=50025
before=50025, after=50763
```

<details>
<summary>measure.py</summary>

```python
import gc
import sys

for _ in range(5):
    gc.collect()
    before = sys.gettotalrefcount()

    import _socket
    del sys.modules["_socket"]
    del _socket

    gc.collect()
    after = sys.gettotalrefcount()

    print(f"{before=}, {after=}")
    # assert after == before
```

</details>

... and these two:

```console
$ ./python.exe -m test -R : test_socket -m RecvmsgIntoSCMRightsStreamTest                                                                              Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 1.96 Run tests sequentially
0:00:00 load avg: 1.96 [1/1] test_socket
beginning 9 repetitions
123456789
.........
test_socket leaked [10, 10, 10, 10] file descriptors, sum=40
test_socket failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_socket

Total duration: 1.8 sec
Tests result: FAILURE
$ ./python.exe -m test -R : test_socket -m RecvmsgSCMRightsStreamTest                                                                                  Raised RLIMIT_NOFILE: 256 -> 1024
0:00:00 load avg: 2.52 Run tests sequentially
0:00:00 load avg: 2.52 [1/1] test_socket
beginning 9 repetitions
123456789
.........
test_socket leaked [10, 10, 10, 10] file descriptors, sum=40
test_socket failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_socket

Total duration: 1.8 sec
Tests result: FAILURE
```

**Comment by erlend-aasland** [comment]
Regarding https://github.com/python/cpython/pull/103094#issuecomment-1495971330: the file descriptors leak on `main` as well, so it is not a regression introduced by this PR. Let's not bother with those. There's still references leaked by repeated init of the module, though.

**Comment by erlend-aasland** [comment]
I'd like to hold this PR until #103261 has landed.

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Pull request merged**
